### PR TITLE
feat: Show team user availability (WEBAPP-4517, WEBAPP-4518, WEBAPP-4521)

### PR DESCRIPTION
**Pull request by gregor**

### CODE REVIEW

File: app/page/template/content/preferences-account.htm [hunk 1]
```diff
@@ -13,8 +13,14 @@
                    spellcheck="false" />
             <span class="preferences-account-input-icon" data-bind="css: name_saved() ? 'checkmark' : 'preferences-account-input-icon-edit'"></span>
           </div>
+          <!-- ko if: is_team() -->
+            <availability-state class="preferences-account-availability"
+                                data-bind="click: clickOnAvailability"
+                                data-uie-name="status-availability-in-profile"
+                                params="availability: availability, label: availabilityLabel, showArrow: true"></availability-state>
+          <!-- /ko -->
```
Comment (lipis) [inline app/page/template/content/preferences-account.htm]: If somebody could build the html parser for Prettier :D

File: app/script/entity/Conversation.js [hunk 2]
```diff
@@ -394,19 +394,15 @@ z.entity.Conversation = class Conversation {
...
-        return userEt.devices().length;
+        return z.client.ClientRepository.CONFIG.AVERAGE_NUMBER_OF_CLIENTS;
       }, this.self.devices().length);
-
-      const estimatedUnknownClients = usersWithoutKnownClients * 4;
-      return numberOfKnownClients + estimatedUnknownClients;
     }
 
-    return this.get_number_of_participants() * 4;
+    return this.get_number_of_participants() * z.client.ClientRepository.CONFIG.AVERAGE_NUMBER_OF_CLIENTS;
```
Comment (lipis) [inline app/script/entity/Conversation.js]: life is so much nicer without the magic numbers.. I doubt if anyone would know what is this 4 in a year from now..

Comment (lipis) [reply app/script/entity/Conversation.js]: if not in a month :D

File: app/script/user/AvailabilityMapper.js [hunk 3]
```diff
@@ -37,6 +37,20 @@ z.user.AvailbilityMapper = (() => {
...
+    switch (availabilityType) {
+      case z.user.AvailabilityType.AVAILABLE:
+        return z.proto.Availability.Type.AVAILABLE;
+      case z.user.AvailabilityType.AWAY:
+        return z.proto.Availability.Type.AWAY;
+      case z.user.AvailabilityType.BUSY:
+        return z.proto.Availability.Type.BUSY;
+      case z.user.AvailabilityType.NONE:
+        return z.proto.Availability.Type.NONE;
+      default:
```
Comment (lipis) [inline app/script/user/AvailabilityMapper.js:51]: shouldn't be the `default` z.proto.Availability.Type.NONE.. so maybe replacing the line `48` with the `default`

Comment (gregor) [reply app/script/user/AvailabilityMapper.js:51]: No, I want to be explicit. Added throwing errors for the default cases instead.

File: app/page/template/partials/template-user-profile.htm [hunk 4]
```diff
@@ -33,6 +33,11 @@
               <!-- /ko -->
             </div>
           <!-- /ko -->
+          <!-- if: is_team -->
+            <availability-state class="user-profile-details-availability"
+                                data-uie-name="status-availability"
+                                params="availability: user().availability(), label: userAvailabilityLabel()"></availability-state>
```
Comment (bennycode) [inline app/page/template/partials/template-user-profile.htm]: Needs lipification. 😀

File: app/script/components/availabilityState.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+z.components.AvailabilityState = class AvailabilityState {
+  constructor(params) {
+    this.availability = params.availability;
+    this.label = params.label;
+    this.showArrow = params.showArrow || false;
+    this.theme = params.theme || false;
+  }
+};
+
+// Knockout registration of the availability state component.
```
Comment (bennycode) [inline app/script/components/availabilityState.js]: No need for this comment.

File: app/script/components/userProfile.js [hunk 6]
```diff
@@ -47,6 +47,18 @@ z.components.UserProfileViewModel = class UserProfileViewModel {
...
 
+    this.is_team = ko.pureComputed(() => {
+      return this.conversation()
+        .self()
+        .is_team_member();
+    });
+    this.userAvailabilityLabel = ko.pureComputed(() => {
+      const availabilitySetToNone = this.user().availability() === z.user.AvailabilityType.NONE;
+      if (!availabilitySetToNone) {
+        return z.user.AvailbilityMapper.nameFromType(this.user().availability());
```
Comment (bennycode) [inline app/script/components/userProfile.js]: `AvailbilityMapper` -> `AvailabilityMapper`

File: app/script/conversation/ClientMismatchHandler.js [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+    this.logger.debug(`Message contains redundant clients of '${Object.keys(recipients).length}' users`, recipients);
+
+    return this.conversationRepository
+      .get_conversation_by_id(conversationId)
+      .catch(error => {
+        if (error.type !== z.conversation.ConversationError.TYPE.NOT_FOUND) {
+          throw error;
+        }
+      })
+      .then(conversationEt => {
```
Comment (bennycode) [inline app/script/conversation/ClientMismatchHandler.js]: I suggest to skip the `conversation_et` and `conversation_ets` pattern: `conversationEt` -> `conversationEntity`

File: app/script/entity/User.js [hunk 8]
```diff
@@ -155,14 +155,20 @@ z.entity.User = class User {
 
     this.is_request = ko.pureComputed(() => this.connection().is_request());
 
-    // e2ee
+    // E2EE
```
Comment (bennycode) [inline app/script/entity/User.js]: We can remove this comment, it doesn't give much information anymore.

File: test/unit_tests/conversation/ClientMismatchHandlerSpec.js [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+        .onClientMismatch(clientMismatch, conversationEt.id, genericMessage, payload)
+        .then(updatedPayload => {
+          expect(Object.keys(updatedPayload.recipients).length).toBe(2);
+          expect(Object.keys(updatedPayload.recipients[johnDoe.user_id]).length).toBe(1);
+          done();
+        })
+        .catch(done.fail);
+    });
+
+    xit('should remove the payload of deleted clients', done => {
```
Comment (bennycode) [inline test/unit_tests/conversation/ClientMismatchHandlerSpec.js]: Should this test case be removed?

File: test/unit_tests/cryptography/CryptographyMapperSpec.js [hunk 10]
```diff
@@ -224,6 +224,27 @@ describe('z.cryptography.CryptographyMapper', () => {
         .catch(done.fail);
     });
 
+    it('resolves with a mapped availabilitymessage', done => {
```
Comment (bennycode) [inline test/unit_tests/cryptography/CryptographyMapperSpec.js]: `availabilitymessage` -> `availability message`

File: app/script/ui/AvailabilityContextMenu.js [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ *
+ */
+
+'use strict';
+
+window.z = window.z || {};
+window.z.ui = z.ui || {};
+
+z.ui.AvailibilityContextMenu = (() => {
```
Comment (bennycode) [inline app/script/ui/AvailabilityContextMenu.js]: `z.ui.AvailibilityContextMenu` -> `z.ui.AvailabilityContextMenu`

**Pull request merged**